syntax = "proto2";

import "google/protobuf/descriptor.proto";

package validate;

// Validation rules for nanopb
message FieldRules {
    // String rules
    message StringRules {
        optional uint64 min_len = 1;
        optional uint64 max_len = 2;
        optional string contains = 3;
        optional string prefix = 4;
        optional string suffix = 5;
    }
    optional StringRules string = 1;
    
    // Integer rules
    message Int32Rules {
        optional int32 gte = 1;
        optional int32 lte = 2;
        optional int32 gt = 3;
        optional int32 lt = 4;
        repeated int32 in = 5;
        repeated int32 not_in = 6;
    }
    optional Int32Rules int32 = 2;
    
    // Float rules
    message FloatRules {
        optional float gte = 1;
        optional float lte = 2;
        optional float gt = 3;
        optional float lt = 4;
    }
    optional FloatRules float = 3;
    
    // Enum rules
    message EnumRules {
        optional bool defined_only = 1;
    }
    optional EnumRules enum = 4;
    
    // Repeated field rules
    message RepeatedRules {
        optional uint64 min_items = 1;
        optional uint64 max_items = 2;
        optional bool unique = 3;
    }
    optional RepeatedRules repeated = 5;
    
    // General rules
    optional bool required = 6;
}

// Field validation rules
extend google.protobuf.FieldOptions {
    optional FieldRules rules = 1001;
}

// File-level validation enablement
extend google.protobuf.FileOptions {
    optional bool validate = 1001;
}
